### PR TITLE
Implemented Struck SIS3316 DAQ (via ORCA) decoder

### PR DESCRIPTION
**Pull request by jbrowni2**
Before submitting a pull request, please make sure you've read and understood the pygama developer's guide: https://legend-exp.github.io/pygama/main/developer.html. In particular, do not forget to:

- [ ] Follow our **coding conventions**
- [ ] Update existing or add new **tests**
- [ ] Update existing or add new **documentation**
- [ ] Address any issue reported by GitHub bots

### CODE REVIEW

**Comment by mmatteo** [comment]
@jbrowni2, who are you?

**Comment by jbrowni2** [comment]
My name is James Browning. I am a graduate student for Dr. Green working on COHERENT at North Carolina State University.

**Comment by gipert** [comment]
Hi James, I would recommend to update your GitHub profile with your name and institution.

Thanks for the contribution! Before it can be merged into pygama, please have a look at our [developer's guide](https://legend-exp.github.io/pygama/main/developer.html) and make sure your code conforms to our standards. Please also fix all [issues reported by pre-commit](https://results.pre-commit.ci/run/github/151981074/1659611645.p42dV8OKQbmLhSQuy1LY0Q).

**Comment by jbrowni2** [comment]
Sure thing.

**Comment by jasondet** [comment]
Thanks James. Please do the following:
- Fix all of the [pre-commit.ci](https://results.pre-commit.ci/run/github/151981074/1659618832.q2SA1IvQQHOa4tlOgz8yCQ) errors
- get_key(), get_fcid(), and get_ch() are not needed and should be removed
- self.decoder_name, self.orca_class_name, and self.ievt are not needed and should be removed
- raise exceptions instead of printing error statements
- get_key_list() needs to return the list of valid keys
- assert_nsamples() is not needed and should be removed
- evt_data_32 is not needed because `packet` is already a uint32 buffer
- evt_data_16 should use frombuffer() instead of fromstring() (did this even work??)
- the decoding looks odd. For example, crate/card/channel should all come from packet[1] and wf_samples from a different word. Please make sure to test the code, let us know if you have verified the decoding.

See the SIS3302 code in orca_digitizers.py which serves as a better reference / starting point for this code than the flashcams decoder.

**Comment by jbrowni2** [comment]
I am working on making the changes now. I more or less just copied the decoder from the previous version of pygama.

I did test that this decoder works with this version of pygama.

**Comment by jbrowni2** [comment]
in pre-commit.ci I keep failing at code spell, but all of the others seem to work. Can you check my code to let me know what I need to change.

File: src/pygama/raw/orca/orca_SIS3316.py [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+            "card": {
+                "dtype": "uint8",
+            },
+            "channel": {
+                "dtype": "uint8",
+            },
+            "waveform": {
+                "dtype": "uint16",
+                "datatype": "waveform",
+                "wf_len": 65532,  # max value. override this before initalizing buffers to save RAM
```
Comment (gipert) [inline src/pygama/raw/orca/orca_SIS3316.py:51]: initalizing -> initializing

File: src/pygama/raw/orca/orca_SIS3316.py [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+
+log = logging.getLogger(__name__)
+
+
+class ORSIS3316WaveformDecoder(OrcaDecoder):
+    """Decoder for FlashCam ADC data written by ORCA."""
+
+    def __init__(self, header: OrcaHeader = None, **kwargs) -> None:
+        # self.decoder_name = "ORSIS3316WaveformDecoder"
+        # self.orca_class_name = "ORSIS3316Model"
```
Comment (gipert) [inline src/pygama/raw/orca/orca_SIS3316.py:21]: Please remove all commented code, leave only comments documenting the code. Please fix this also below.

File: src/pygama/raw/orca/orca_SIS3316.py [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+                    trace_length = card_dict["rawDataBufferLen"]
+                    self.decoded_values[ccc] = copy.deepcopy(
+                        self.decoded_values_template
+                    )
+
+                    if trace_length <= 0 or trace_length > 2**16:
+                        raise Exception(
+                            "SIS3316ORCADecoder Error: invalid trace_length",
+                            trace_length,
+                        )
```
Comment (gipert) [inline src/pygama/raw/orca/orca_SIS3316.py:88]:
```suggestion
                        raise RuntimeError("invalid trace_length: ", trace_length)
```
Please raise this if something unexpected happens during decoding at runtime, `Exception` is too generic. `SIS3316ORCADecoder Error: ` is not needed since it will be clear from the stacktrace. Please fix this also in the rest of the code.

File: src/pygama/raw/orca/orca_streamer.py [hunk 4]
```diff
@@ -9,12 +9,12 @@
 from pygama.raw.data_streamer import DataStreamer
 from pygama.raw.orca import orca_packet
 from pygama.raw.orca.orca_base import OrcaDecoder
-from pygama.raw.orca.orca_digitizers import ORSIS3302DecoderForEnergy  # noqa: F401
```
Comment (gipert) [inline src/pygama/raw/orca/orca_streamer.py:12]: Why did you remove this line?

File: src/pygama/raw/orca/orca_SIS3316.py [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+from pygama.raw.orca.orca_packet import OrcaPacket
+from pygama.raw.raw_buffer import RawBufferLibrary
+
+from .orca_base import OrcaDecoder, get_ccc
+
+log = logging.getLogger(__name__)
+
+
+class ORSIS3316WaveformDecoder(OrcaDecoder):
+    """Decoder for FlashCam ADC data written by ORCA."""
```
Comment (gipert) [inline src/pygama/raw/orca/orca_SIS3316.py:17]: Please update docstrings for all classes and methods with correct and complete documentation. Have a look at [our guide](https://legend-exp.github.io/pygama/releases/v1.0/developer.html#documentation) for how to write and build the documentation locally. You can also have a look at how other docstrings are formatted in pygama.

**Review by gipert — CHANGES_REQUESTED** [review]
I've left a couple of comments inline, I'll leave to @jasondet a more accurate review of what the code is specifically doing.

Do you have a small test file (a few MB) that we can upload to https://github.com/legend-exp/legend-testdata for automatic testing?

**Comment by jbrowni2** [comment]
I have a file that is 200 mb is that to large?

**Comment by gipert** [comment]
Yes, but you can shorten it with `dd` so to have a few tens of event file. For example: `dd if=input.orca of=output.orca bs=1M count=s`. Try to cut the file at the right byte if you can...

**Comment by jbrowni2** [comment]
I have a small file for testing. How am I supposed to add it to lend-exp/legend-testdata? would you like me to email it to you?

**Comment by gipert** [comment]
Yes, please email me at luigi.pertoldi@tum.de.

**Review by gipert — APPROVED** [review]

**Comment by gipert** [comment]
@jasondet if changes look fine to you, I will squash all this and merge.

File: src/pygama/raw/orca/orca_digitizers.py [hunk 6]
```diff
@@ -210,3 +204,228 @@ def decode_packet(self, packet, packet_id, rbl):
...
+        self.decoded_values = {}
+        self.skipped_channels = {}
+        super().__init__(
+            header=header, **kwargs
+        )  # also initializes the garbage df (whatever that means...)
+
+    def set_header(self, header: OrcaHeader) -> None:
+        self.header = header
+
+        self.decoded_values = copy.deepcopy(self.decoded_values_template)
```
Comment (jasondet) [inline src/pygama/raw/orca/orca_digitizers.py:260]: this line seems like a bug. It will create a dec vals dict at self.decoded_values that is later appended at self.decoded_values[ccc]

I think this line can just be removed?

File: src/pygama/raw/orca/orca_digitizers.py [hunk 7]
```diff
@@ -210,3 +204,228 @@ def decode_packet(self, packet, packet_id, rbl):
...
+
+        for card_dict in self.header["ObjectInfo"]["Crates"][0]["Cards"]:
+            if card_dict["Class Name"] == "ORSIS3316Model":
+                card = card_dict["Card"]
+                crate = 0
+                for channel in range(0, 16):
```
Comment (jasondet) [inline src/pygama/raw/orca/orca_digitizers.py:266]: This loop is missing a check for which channels are enabled.

File: src/pygama/raw/orca/orca_digitizers.py [hunk 8]
```diff
@@ -210,3 +204,228 @@ def decode_packet(self, packet, packet_id, rbl):
...
+
+    def get_key_list(self) -> list[int]:
+        key_list = []
+        for key in self.decoded_values.keys():
+            key_list += [key]
+        return key_list
+
+    def get_decoded_values(self, key: int = None) -> dict[str, Any]:
+        if key is None:
+            dec_vals_list = self.decoded_values
```
Comment (jasondet) [inline src/pygama/raw/orca/orca_digitizers.py:291]: note: dec_vals_list is a dict, not a list...

I suggest setting it to self.decoded_values.values()

File: src/pygama/raw/orca/orca_digitizers.py [hunk 9]
```diff
@@ -210,3 +204,228 @@ def decode_packet(self, packet, packet_id, rbl):
...
+            if len(dec_vals_list) == 0:
+                raise RuntimeError("decoded_values not built yet!")
+
+            return dec_vals_list  # Get first thing we find
+
+        if key in self.decoded_values:
+            dec_vals_list = self.decoded_values[key]
+            return dec_vals_list
```
Comment (jasondet) [inline src/pygama/raw/orca/orca_digitizers.py:299]: the previous two lines can be combined

File: src/pygama/raw/orca/orca_digitizers.py [hunk 10]
```diff
@@ -210,3 +204,228 @@ def decode_packet(self, packet, packet_id, rbl):
...
+            key_list += [key]
+        return key_list
+
+    def get_decoded_values(self, key: int = None) -> dict[str, Any]:
+        if key is None:
+            dec_vals_list = self.decoded_values
+            if len(dec_vals_list) == 0:
+                raise RuntimeError("decoded_values not built yet!")
+
+            return dec_vals_list  # Get first thing we find
```
Comment (jasondet) [inline src/pygama/raw/orca/orca_digitizers.py:295]: this line will need to be edited when my earlier comments are addressed

File: src/pygama/raw/orca/orca_digitizers.py [hunk 11]
```diff
@@ -210,3 +204,228 @@ def decode_packet(self, packet, packet_id, rbl):
...
+            return dec_vals_list
+        raise RuntimeError("no decoded values for key", key)
+
+    def decode_packet(
+        self, packet: OrcaPacket, packet_id: int, rbl: RawBufferLibrary
+    ) -> bool:
+        """Decode the ORCA SIS3316 ADC packet.
+
+        The packet is formatted as:
+
```
Comment (jasondet) [inline src/pygama/raw/orca/orca_digitizers.py:308]: I couldn't find this card documented at
http://orca.physics.unc.edu/orca/VME/Index.html
Let's ask @tcald to add a page on it.

Comment (jbrowni2) [reply src/pygama/raw/orca/orca_digitizers.py:371]: I added those in my new commit along with fixing the above issues.

**Comment by gipert** [comment]
@jbrowni2 could you please mark resolved conversations as "resolved"?

**Review by jasondet — APPROVED** [review]

**Comment by jasondet** [comment]
this looks good now, thanks for pushing it through the long process @jbrowni2 !

**Comment by jbrowni2** [comment]
I am James Browning. I am a graduate student for Dr. Green working on
COHERENT.
